File: packages/utils/src/algorithms/trie.ts
/**
 * Trie (prefix tree) for O(1) average-case member prefix search.
 * Indexes member names, emails, and membership IDs for fast autocomplete.
 */

interface TrieNode {
  children: Map<string, TrieNode>;
  memberIds: Set<string>;
  isTerminal: boolean;
}

export class MemberSearchTrie {
  private root: TrieNode = this.createNode();

  private createNode(): TrieNode {
    return { children: new Map(), memberIds: new Set(), isTerminal: false };
  }

  private normalize(text: string): string {
    return text.toLowerCase().trim();
  }

  insert(memberId: string, ...fields: string[]): void {
    for (const field of fields) {
      const normalized = this.normalize(field);
      let node = this.root;

      for (const char of normalized) {
        if (!node.children.has(char)) {
          node.children.set(char, this.createNode());
        }
        node = node.children.get(char)!;
        node.memberIds.add(memberId);
      }
      node.isTerminal = true;
    }
  }

  remove(memberId: string, ...fields: string[]): void {
    for (const field of fields) {
      const normalized = this.normalize(field);
      let node = this.root;

      for (const char of normalized) {
        const child = node.children.get(char);
        if (!child) break;
        child.memberIds.delete(memberId);
        node = child;
      }
    }
  }

  search(prefix: string, limit = 50): string[] {
    const normalized = this.normalize(prefix);
    let node = this.root;

    for (const char of normalized) {
      const child = node.children.get(char);
      if (!child) return [];
      node = child;
    }

    return [...node.memberIds].slice(0, limit);
  }

  clear(): void {
    this.root = this.createNode();
  }

  size(): number {
    return this.root.memberIds.size;
  }
}

File: apps/api/v1/core/algorithms/trie.py
"""Trie (prefix tree) for fast member prefix search."""


class MemberSearchTrie:
    def __init__(self) -> None:
        self._root: dict = {"children": {}, "member_ids": set(), "terminal": False}

    def _create_node(self) -> dict:
        return {"children": {}, "member_ids": set(), "terminal": False}

    @staticmethod
    def _normalize(text: str) -> str:
        return text.lower().strip()

    def insert(self, member_id: str, *fields: str) -> None:
        for field in fields:
            normalized = self._normalize(field)
            node = self._root
            for char in normalized:
                if char not in node["children"]:
                    node["children"][char] = self._create_node()
                node = node["children"][char]
                node["member_ids"].add(member_id)
            node["terminal"] = True

    def remove(self, member_id: str, *fields: str) -> None:
        for field in fields:
            normalized = self._normalize(field)
            node = self._root
            for char in normalized:
                child = node["children"].get(char)
                if not child:
                    break
                child["member_ids"].discard(member_id)
                node = child

    def search(self, prefix: str, limit: int = 50) -> list[str]:
        normalized = self._normalize(prefix)
        node = self._root
        for char in normalized:
            child = node["children"].get(char)
            if not child:
                return []
            node = child
        return list(node["member_ids"])[:limit]

    def clear(self) -> None:
        self._root = self._create_node()
